### PR TITLE
New Parser (handle arbitrary amounts of indents/dedents, better error handling, handles quotes, more tests, output is nested arrays)

### PR DESCRIPTION
**Pull request by melaniecebula**
New features:
- handles arbitrary amounts of nesting (indents/dedents)
- output is nested JSON arrays to better capture nesting
- better error handling for inconsistent indentation and JSON.parse failure
- handles quotes
- a lot of tests added

May want to update README to show examples of nested input/output

Issue: https://github.com/mafintosh/hackfile/issues/4

### CODE REVIEW

**Comment by melaniecebula** [comment]
This PR is a bit of a monster, so let me know if you'd rather I break it up into smaller PRs

**Comment by karissa** [comment]
woot!

**Comment by melaniecebula** [comment]
success!

**Comment by mafintosh** [comment]
the indentation seems a bit weird.

**Comment by melaniecebula** [comment]
Can you explain further?

**Comment by mafintosh** [comment]
@melaniecebula i mean the indentation in the source code of your PR seems off. its hard to see where the main function ends

**Comment by melaniecebula** [comment]
oof, you're right.  Looks like I need to fix my vim's auto-indent :/ .  Updated (hopefully correct now but let me know)

**Comment by mafintosh** [comment]
It still seems a bit off around line 60 and until the end of the function

**Comment by mafintosh** [comment]
Instead of the `toJSON` method we could do something like this

``` js
var map = function(commands) {
  var result = []
  while (commands.length) {
    var token = commands.shift()
    if (token === 'INDENT') {
      result.push(map(commands))
      continue
    }
    if (token === 'DEDENT') {
      return result
    }
    result.push(token)
  }
  return result
}

var hackfile = function(src) {
  ...
  return commands.map(map)
}
```

btw your `INDENT`, `DEDENT` parser seems super solid. :)

**Comment by melaniecebula** [comment]
I like the map method a lot, but I'm not sure how to handle nesting nicely with it.  For example, currently test indent-dedent fails because it doesn't put nested commands in a nested array:

```
    expected:
      [ [ 'pipeline', [ 'foo', [ 'pipe', [ 'echo hello', 'transform', 'cat' ] ], [ 'run', [ 'echo', 'hello' ] ] ] ] ]
    actual:
      [ [ 'pipeline', [ 'foo', 'pipe', [ 'echo hello', 'transform', 'cat' ], 'run', [ 'echo', 'hello' ] ] ] ]
```
Is there a good way to handle this that you can think of?

**Comment by mafintosh** [comment]
@melaniecebula try to see if this quick hack fixes it :)

``` js
var map = function(commands) {
  var result = []
  while (commands.length) {
    var token = commands.shift()
    if (token === 'INDENT') {
      result.push([result.pop(), map(commands)])
      continue
    }
    if (token === 'DEDENT') {
      return result
    }
    result.push(token)
  }
  return result
}

var hackfile = function(src) {
  ...
  return commands.map(function(cmd) {
    return map(cmd)[0]
  })
}
```

**Comment by karissa** [comment]
i was thinking about this when first looking at the lower-level format.. does it make sense to have commands be represented as json instead of as lists? is there any particular reason they are lists?

**Comment by mafintosh** [comment]
@karissa ordering :) we could remove the top level array though and turn it into something like:

``` js
[{
  name:'pipe',
  params:[
    'echo 1',
    'echo1', 
    {
      name:'pipe', ...
      params: [...]
    }
}]
```

**Comment by mafintosh** [comment]
i don't have any strong opinions either way

**Comment by melaniecebula** [comment]
@mafintosh:  that did the trick, awesome job!  I've pushed the latest changes.

**Comment by melaniecebula** [comment]
@karissa :  Yeah, we originally had objects until we decided that because of nesting we needed lists.  But I'm open to tweaks to the output

**Comment by karissa** [comment]
i think it might be smart to have objects because you it is more flexible

**Comment by mafintosh** [comment]
@karissa does the `[{name:.., params:[...]}]` syntax look ok?

**Comment by karissa** [comment]
yeah, i think its definitely better. what's `echo` and `echo1` in those examples? references to other pipelines/tasks?

**Comment by mafintosh** [comment]
@karissa its unnested commands. like this example parsed

```
pipe
  echo 1
  echo 2
  pipe
    subcommands
```

**Comment by karissa** [comment]
so the only things that will be objects will have subcommands?

File: index.js [hunk 1]
```diff
@@ -8,26 +10,117 @@ var hackfile = function(src) {
...
+  debug("result before attempting to parse", result)
+  try {
+    result = JSON.parse(result)
+    debug("JSON parsed res", result)
+    return result
+  }catch(e) {
+    console.log("JSON could not parse " + result) 
+    debug("JSON parse error", e)
+    return
+  }
```
Comment (mafintosh) [inline index.js]: lines 113-122 aren't really needed anymore

File: index.js [hunk 2]
```diff
@@ -8,26 +10,117 @@ var hackfile = function(src) {
...
+  numTimes = numIndents/indentSize
+  for (var i = 0; i < numTimes; i += 1) array.push(token)
+}
+
+/* Trims and handles quotes in INPUT */
+var sanitize = function(input) {
+  var SINGLE_QUOTE = /\'/g
+  var DOUBLE_QUOTE = /\"/g
+  return input.trim().replace(DOUBLE_QUOTE, '\\"').replace(SINGLE_QUOTE, '\'')
+}
```
Comment (mafintosh) [inline index.js]: pretty sure we don't need the sanitize function anymore as we don't json.parse/stringify

**Comment by mafintosh** [comment]
@karissa yea i guess since thats the "primitives"

**Comment by melaniecebula** [comment]
@mafintosh removed JSON.parse code, sanitize code

**Comment by maxogden** [comment]
IMO output from this parser should be as low level as possible. When it gets converted into gasket-compatible JSON then we can worry about human readability/flexibility etc

**Comment by maxogden** [comment]
made a thing with gasket today https://github.com/maxogden/npm-readme-search. see README for instructions, and package.json for the definitions. Just posting it here because it's a real use case that might be interesting to think about how it would work with hackfile/datscript.

One thing in particular that would be cool is a way to say 'run these two processes as long running child processes and exit both if my process exits'. The way I do it now is `"start": "node foo.js & node bar.js"` but that results in zombie processes for me quite often because of the way `&` works.

**Comment by maxogden** [comment]
also if you follow the instructions in that readme you can see it totally relies on the commands themselves to provide output to the user. some don't have any progress/messages output at all. maybe theres something we can do to make user-friendly progress output by default? (this is outside the scope of hackfiles probably, but im doing a braindump right now hehe)

**Comment by mafintosh** [comment]
Released in 3.0.0

**Pull request merged**